File: gitwork.py
#!/usr/bin/env python3
"""
gitwork - Git commit analyzer with time shortcuts and multiple output formats
"""

import argparse
import csv
import json
import os
import subprocess
import sys
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Optional


class Commit:
    def __init__(self, hash: str, author: str, email: str, date: str, subject: str, body: str = ""):
        self.hash = hash[:7]
        self.author = author
        self.email = email
        self.date = date
        self.subject = subject
        self.body = body.strip()

    def to_dict(self):
        d = {
            "hash": self.hash,
            "author": self.author,
            "email": self.email,
            "date": self.date,
            "subject": self.subject,
        }
        if self.body:
            d["body"] = self.body
        return d


class GitWork:
    def __init__(self):
        self.repo_dir = "."
        self.since = None
        self.until = None
        self.authors = []
        self.all_authors = False
        self.mode = "pretty"
        self.merges = ""
        self.file_path = ""
        self.grep_patterns = []
        self.explicit_author = False

    def parse_args(self):
        parser = argparse.ArgumentParser(
            description="Git commit analyzer",
            formatter_class=argparse.RawDescriptionHelpFormatter,
            epilog="""
Time shortcuts:
  today, yesterday, week, month, monday, 3d, 7d, 2w, 1m

Examples:
  gitwork.py today
  gitwork.py week --authors alice@co.com
  gitwork.py month --json --all
  gitwork.py week --path ~/projects/api
            """,
        )

        parser.add_argument("time", nargs="?", help="Time shortcut (today, week, 3d, etc)")
        parser.add_argument("--since", help="Since date/time")
        parser.add_argument("--until", help="Until date/time")
        parser.add_argument("--authors", help="Comma-separated author emails")
        parser.add_argument("--me", action="store_true", help="Include yourself")
        parser.add_argument("--all", action="store_true", dest="all_authors", help="Show all authors")
        parser.add_argument("--oneline", action="store_const", const="oneline", dest="mode")
        parser.add_argument("--full", action="store_const", const="full", dest="mode")
        parser.add_argument("--stat", action="store_const", const="stat", dest="mode")
        parser.add_argument("--diff", action="store_const", const="diff", dest="mode")
        parser.add_argument("--files", action="store_const", const="files", dest="mode")
        parser.add_argument("--summary", action="store_true", help="Show summary")
        parser.add_argument("--json", action="store_true", help="JSON output")
        parser.add_argument("--csv", action="store_true", help="CSV output")
        parser.add_argument("--no-merges", action="store_const", const="--no-merges", dest="merges")
        parser.add_argument("--only-merges", action="store_const", const="--merges", dest="merges")
        parser.add_argument("--path", help="Repository directory")
        parser.add_argument("--file-path", help="Limit to subdirectory within repo")
        parser.add_argument("--grep", action="append", dest="grep_patterns", help="Filter by message")

        args = parser.parse_args()

        # Parse time shortcut
        if args.time:
            self.since = self.parse_time_shortcut(args.time)
        if args.since:
            self.since = args.since
        if args.until:
            self.until = args.until

        # Authors
        if args.authors:
            self.explicit_author = True
            self.authors = [a.strip() for a in args.authors.split(",") if a.strip()]
        if args.me:
            self.explicit_author = True
            me = self.get_git_config("user.email")
            if me:
                self.authors.append(me)
        if args.all_authors:
            self.explicit_author = True
            self.authors = []

        # Mode
        if args.mode:
            self.mode = args.mode
        if args.json:
            self.mode = "json"
        if args.csv:
            self.mode = "csv"
        if args.summary:
            self.mode = "summary"

        # Filters
        if args.merges:
            self.merges = args.merges
        if args.file_path:
            self.file_path = args.file_path
        if args.grep_patterns:
            self.grep_patterns = args.grep_patterns

        # Repo path
        if args.path:
            self.repo_dir = os.path.abspath(args.path)
        else:
            self.repo_dir = os.getcwd()

    def parse_time_shortcut(self, shortcut: str) -> str:
        shortcuts = {
            "today": "midnight",
            "yesterday": "yesterday",
            "week": "1 week ago",
            "month": "1 month ago",
            "monday": "last Monday",
        }
        if shortcut in shortcuts:
            return shortcuts[shortcut]
        if shortcut.endswith("d"):
            return f"{shortcut[:-1]} days ago"
        if shortcut.endswith("w"):
            return f"{shortcut[:-1]} weeks ago"
        if shortcut.endswith("m"):
            return f"{shortcut[:-1]} months ago"
        return shortcut

    def validate_repo(self):
        try:
            subprocess.run(
                ["git", "-C", self.repo_dir, "rev-parse", "--is-inside-work-tree"],
                check=True,
                capture_output=True,
            )
        except subprocess.CalledProcessError:
            print(f"❌ Not a git repository: {self.repo_dir}", file=sys.stderr)
            sys.exit(1)

    def get_git_config(self, key: str) -> Optional[str]:
        try:
            result = subprocess.run(
                ["git", "-C", self.repo_dir, "config", key],
                capture_output=True,
                text=True,
                check=True,
            )
            return result.stdout.strip()
        except subprocess.CalledProcessError:
            return None

    def fetch_commits(self) -> List[Commit]:
        cmd = ["git", "-C", self.repo_dir, "log", "--pretty=format:%H\x1F%an\x1F%ae\x1F%aI\x1F%s\x1F%b\x1E"]

        if self.since:
            cmd.append(f"--since={self.since}")
        if self.until:
            cmd.append(f"--until={self.until}")
        if self.merges:
            cmd.append(self.merges)
        for author in self.authors:
            cmd.append(f"--author={author}")
        for pattern in self.grep_patterns:
            cmd.append(f"--grep={pattern}")
        if self.file_path:
            cmd.extend(["--", self.file_path])

        try:
            result = subprocess.run(cmd, capture_output=True, text=True, check=True)
        except subprocess.CalledProcessError:
            return []

        commits = []
        for record in result.stdout.split("\x1E"):
            record = record.strip()
            if not record:
                continue
            fields = record.split("\x1F")
            if len(fields) < 5:
                continue
            commit = Commit(
                hash=fields[0],
                author=fields[1],
                email=fields[2],
                date=fields[3],
                subject=fields[4],
                body=fields[5] if len(fields) > 5 else "",
            )
            commits.append(commit)
        return commits

    def output_json(self, commits: List[Commit]):
        data = [c.to_dict() for c in commits]
        print(json.dumps(data, indent=2))

    def output_csv(self, commits: List[Commit]):
        writer = csv.writer(sys.stdout)
        writer.writerow(["hash", "author", "date", "subject"])
        for c in commits:
            writer.writerow([c.hash, c.author, c.date, c.subject])

    def output_summary(self, commits: List[Commit]):
        print(f"{'Total commits:':<18} {len(commits)}")

    def output_pretty(self, commits: List[Commit]):
        prev_date = None
        for c in commits:
            date = c.date[:10]
            if date != prev_date:
                if prev_date:
                    print()
                print(f"\n📅 \033[1;36m{date}\033[0m")
                prev_date = date
            print(f"  • \033[1;33m{c.hash}\033[0m {c.subject}")
            if c.body:
                for line in c.body.split("\n"):
                    line = line.strip()
                    if line:
                        print(f"    \033[90m{line}\033[0m")

    def run(self):
        self.parse_args()
        self.validate_repo()

        # Default to --me if no explicit author
        if not self.explicit_author:
            me = self.get_git_config("user.email")
            if me:
                self.authors = [me]

        commits = self.fetch_commits()

        if self.mode == "json":
            self.output_json(commits)
        elif self.mode == "csv":
            self.output_csv(commits)
        elif self.mode == "summary":
            self.output_summary(commits)
        else:
            self.output_pretty(commits)


if __name__ == "__main__":
    GitWork().run()

File: gitwork.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	// "time"
)

type Commit struct {
	Hash    string `json:"hash"`
	Author  string `json:"author"`
	Email   string `json:"email"`
	Date    string `json:"date"`
	Subject string `json:"subject"`
	Body    string `json:"body,omitempty"`
}

type Config struct {
	since          string
	until          string
	authors        []string
	allAuthors     bool
	mode           string
	merges         string
	repoDir        string
	filePath       string
	grepPatterns   []string
	doSummary      bool
	doJSON         bool
	doCSV          bool
	explicitAuthor bool
}

func main() {
	cfg := parseArgs()

	if err := validateRepo(cfg.repoDir); err != nil {
		fmt.Fprintf(os.Stderr, "❌ %v\n", err)
		os.Exit(1)
	}

	if !cfg.explicitAuthor {
		if email := getGitConfig(cfg.repoDir, "user.email"); email != "" {
			cfg.authors = append(cfg.authors, email)
		}
	}

	if cfg.doSummary {
		showSummary(cfg)
		return
	}

	commits := fetchCommits(cfg)

	if cfg.doJSON {
		outputJSON(commits)
	} else if cfg.doCSV {
		outputCSV(commits)
	} else {
		outputPretty(commits, cfg.mode)
	}
}

func parseArgs() *Config {
	cfg := &Config{
		repoDir: ".",
		mode:    "pretty",
	}

	var authorsStr string
	var help bool

	flag.BoolVar(&help, "help", false, "Show help")
	flag.BoolVar(&help, "h", false, "Show help")
	flag.StringVar(&cfg.since, "since", "", "Since date/time")
	flag.StringVar(&cfg.until, "until", "", "Until date/time")
	flag.StringVar(&authorsStr, "authors", "", "Comma-separated author emails")
	flag.BoolVar(&cfg.allAuthors, "all", false, "Show all authors")
	flag.StringVar(&cfg.mode, "oneline", "", "Oneline mode")
	flag.StringVar(&cfg.mode, "full", "", "Full mode")
	flag.StringVar(&cfg.mode, "stat", "", "Stat mode")
	flag.StringVar(&cfg.mode, "diff", "", "Diff mode")
	flag.StringVar(&cfg.mode, "files", "", "Files mode")
	flag.BoolVar(&cfg.doSummary, "summary", false, "Show summary")
	flag.BoolVar(&cfg.doJSON, "json", false, "JSON output")
	flag.BoolVar(&cfg.doCSV, "csv", false, "CSV output")
	flag.StringVar(&cfg.repoDir, "path", ".", "Repository path")
	flag.StringVar(&cfg.filePath, "file-path", "", "File path filter")

	flag.Usage = func() {
		fmt.Println(`Usage: gitwork [time-shortcut] [options]

Time shortcuts:
  today          Since midnight
  yesterday      Since yesterday 00:00
  week           Since 1 week ago
  month          Since 1 month ago
  monday         Since last Monday 00:00
  3d             Since 3 days ago (any <n>d / <n>w / <n>m)

Author options:
  --authors <emails>  Comma-separated author emails
  --all              Show all authors

Output modes:
  --oneline          Compact one line per commit
  --full             Full commit message
  --stat             Show files changed summary
  --diff             Show full diffs
  --files            Only file names
  --summary          Print summary statistics
  --json             Output as JSON
  --csv              Output as CSV

Filters:
  --path <dir>       Repository directory
  --file-path <dir>  Limit to subdirectory within repo

Examples:
  gitwork today
  gitwork week --authors alice@co.com
  gitwork month --json --all
  gitwork week --path ~/projects/api`)
		os.Exit(0)
	}

	flag.Parse()

	if help {
		flag.Usage()
	}

	// Parse time shortcuts from first positional arg
	args := flag.Args()
	if len(args) > 0 {
		cfg.since = parseTimeShortcut(args[0])
	}

	if authorsStr != "" {
		cfg.explicitAuthor = true
		for _, email := range strings.Split(authorsStr, ",") {
			email = strings.TrimSpace(email)
			if email != "" {
				cfg.authors = append(cfg.authors, email)
			}
		}
	}

	if cfg.allAuthors {
		cfg.explicitAuthor = true
		cfg.authors = nil
	}

	// Resolve repo path
	if absPath, err := filepath.Abs(cfg.repoDir); err == nil {
		cfg.repoDir = absPath
	}

	return cfg
}

func parseTimeShortcut(arg string) string {
	switch arg {
	case "today":
		return "midnight"
	case "yesterday":
		return "yesterday"
	case "week":
		return "1 week ago"
	case "month":
		return "1 month ago"
	case "monday":
		return "last Monday"
	default:
		if strings.HasSuffix(arg, "d") {
			return strings.TrimSuffix(arg, "d") + " days ago"
		}
		if strings.HasSuffix(arg, "w") {
			return strings.TrimSuffix(arg, "w") + " weeks ago"
		}
		if strings.HasSuffix(arg, "m") {
			return strings.TrimSuffix(arg, "m") + " months ago"
		}
		return arg
	}
}

func validateRepo(dir string) error {
	cmd := exec.Command("git", "-C", dir, "rev-parse", "--is-inside-work-tree")
	if err := cmd.Run(); err != nil {
		return fmt.Errorf("not a git repository: %s", dir)
	}
	return nil
}

func getGitConfig(dir, key string) string {
	cmd := exec.Command("git", "-C", dir, "config", key)
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func fetchCommits(cfg *Config) []Commit {
	args := []string{"-C", cfg.repoDir, "log", "--pretty=format:%H\x1F%an\x1F%ae\x1F%aI\x1F%s\x1F%b\x1E"}

	if cfg.since != "" {
		args = append(args, "--since="+cfg.since)
	}
	if cfg.until != "" {
		args = append(args, "--until="+cfg.until)
	}
	for _, author := range cfg.authors {
		args = append(args, "--author="+author)
	}
	if cfg.filePath != "" {
		args = append(args, "--", cfg.filePath)
	}

	cmd := exec.Command("git", args...)
	out, err := cmd.Output()
	if err != nil {
		return nil
	}

	var commits []Commit
	records := strings.Split(string(out), "\x1E")
	for _, rec := range records {
		rec = strings.TrimSpace(rec)
		if rec == "" {
			continue
		}
		fields := strings.Split(rec, "\x1F")
		if len(fields) < 5 {
			continue
		}
		commit := Commit{
			Hash:    fields[0][:7],
			Author:  fields[1],
			Email:   fields[2],
			Date:    fields[3],
			Subject: fields[4],
		}
		if len(fields) > 5 {
			commit.Body = strings.TrimSpace(fields[5])
		}
		commits = append(commits, commit)
	}
	return commits
}

func showSummary(cfg *Config) {
	commits := fetchCommits(cfg)
	fmt.Printf("%-18s %d\n", "Total commits:", len(commits))
}

func outputJSON(commits []Commit) {
	data, _ := json.MarshalIndent(commits, "", "  ")
	fmt.Println(string(data))
}

func outputCSV(commits []Commit) {
	w := csv.NewWriter(os.Stdout)
	w.Write([]string{"hash", "author", "date", "subject"})
	for _, c := range commits {
		w.Write([]string{c.Hash, c.Author, c.Date, c.Subject})
	}
	w.Flush()
}

func outputPretty(commits []Commit, mode string) {
	var prevDate string
	for _, c := range commits {
		date := c.Date[:10]
		if date != prevDate {
			if prevDate != "" {
				fmt.Println()
			}
			fmt.Printf("\n📅 \033[1;36m%s\033[0m\n", date)
			prevDate = date
		}
		fmt.Printf("  • \033[1;33m%s\033[0m %s\n", c.Hash, c.Subject)
		if c.Body != "" {
			for _, line := range strings.Split(c.Body, "\n") {
				line = strings.TrimSpace(line)
				if line != "" {
					fmt.Printf("    \033[90m%s\033[0m\n", line)
				}
			}
		}
	}
}
